File: 1.8.9-Forge/src/main/java/net/ccbluex/liquidbounce/features/module/modules/movement/Speed.java
/*
 * LiquidBounce Hacked Client
 * A free open source mixin-based injection hacked client for Minecraft using Minecraft Forge.
 * https://github.com/CCBlueX/LiquidBounce/
 */
package net.ccbluex.liquidbounce.features.module.modules.movement;

import net.ccbluex.liquidbounce.event.*;
import net.ccbluex.liquidbounce.features.module.Module;
import net.ccbluex.liquidbounce.features.module.ModuleCategory;
import net.ccbluex.liquidbounce.features.module.ModuleInfo;
import net.ccbluex.liquidbounce.features.module.modules.movement.speeds.SpeedMode;
import net.ccbluex.liquidbounce.features.module.modules.movement.speeds.aac.*;
import net.ccbluex.liquidbounce.features.module.modules.movement.speeds.ncp.*;
import net.ccbluex.liquidbounce.features.module.modules.movement.speeds.other.*;
import net.ccbluex.liquidbounce.features.module.modules.movement.speeds.spartan.SpartanYPort;
import net.ccbluex.liquidbounce.features.module.modules.movement.speeds.spectre.SpectreBHop;
import net.ccbluex.liquidbounce.features.module.modules.movement.speeds.spectre.SpectreLowHop;
import net.ccbluex.liquidbounce.features.module.modules.movement.speeds.spectre.SpectreOnGround;
import net.ccbluex.liquidbounce.utils.MovementUtils;
import net.ccbluex.liquidbounce.value.BoolValue;
import net.ccbluex.liquidbounce.value.FloatValue;
import net.ccbluex.liquidbounce.value.ListValue;

import java.util.ArrayList;
import java.util.List;

@ModuleInfo(name = "Speed", description = "Allows you to move faster.", category = ModuleCategory.MOVEMENT)
public class Speed extends Module {

    private final SpeedMode[] speedModes = new SpeedMode[] {
            // NCP
            new NCPBHop(),
            new NCPFHop(),
            new SNCPBHop(),
            new NCPHop(),
            new YPort(),
            new YPort2(),
            new NCPYPort(),
            new Boost(),
            new Frame(),
            new MiJump(),
            new OnGround(),

            // AAC
            new AACBHop(),
            new AAC2BHop(),
            new AAC3BHop(),
            new AAC4BHop(),
            new AAC5BHop(),
            new AAC6BHop(),
            new AAC7BHop(),
            new AACHop3313(),
            new AACHop350(),
            new AACLowHop(),
            new AACLowHop2(),
            new AACLowHop3(),
            new AACGround(),
            new AACGround2(),
            new AACYPort(),
            new AACYPort2(),
            new AACPort(),
            new OldAACBHop(),

            // Spartan
            new SpartanYPort(),

            // Spectre
            new SpectreLowHop(),
            new SpectreBHop(),
            new SpectreOnGround(),
            new TeleportCubeCraft(),

            // Server
            new HiveHop(),
            new HypixelHop(),
            new MineplexGround(),

            // Other
            new SlowHop(),
            new CustomSpeed()
    };

    public final ListValue modeValue = new ListValue("Mode", getModes(), "NCPBHop") {

        @Override
        protected void onChange(final String oldValue, final String newValue) {
            if(getState())
                onDisable();
        }

        @Override
        protected void onChanged(final String oldValue, final String newValue) {
            if(getState())
                onEnable();
        }
    };

    public final FloatValue customSpeedValue = new FloatValue("CustomSpeed", 1.6F, 0.2F, 2F);
    public final FloatValue customYValue = new FloatValue("CustomY", 0F, 0F, 4F);
    public final FloatValue customTimerValue = new FloatValue("CustomTimer", 1F, 0.1F, 2F);
    public final BoolValue customStrafeValue = new BoolValue("CustomStrafe", true);
    public final BoolValue resetXZValue = new BoolValue("CustomResetXZ", false);
    public final BoolValue resetYValue = new BoolValue("CustomResetY", false);

    public final FloatValue portMax = new FloatValue("AAC-PortLength", 1, 1, 20);
    public final FloatValue aacGroundTimerValue = new FloatValue("AACGround-Timer", 3F, 1.1F, 10F);
    public final FloatValue cubecraftPortLengthValue = new FloatValue("CubeCraft-PortLength", 1F, 0.1F, 2F);
    public final FloatValue mineplexGroundSpeedValue = new FloatValue("MineplexGround-Speed", 0.5F, 0.1F, 1F);

    @EventTarget
    public void onUpdate(final UpdateEvent event) {
        if(mc.thePlayer.isSneaking())
            return;

        if(MovementUtils.isMoving())
            mc.thePlayer.setSprinting(true);

        final SpeedMode speedMode = getMode();

        if(speedMode != null)
            speedMode.onUpdate();
    }

    @EventTarget
    public void onMotion(final MotionEvent event) {
        if(mc.thePlayer.isSneaking() || event.getEventState() != EventState.PRE)
            return;

        if(MovementUtils.isMoving())
            mc.thePlayer.setSprinting(true);

        final SpeedMode speedMode = getMode();

        if(speedMode != null)
            speedMode.onMotion();
    }

    @EventTarget
    public void onMove(MoveEvent event) {
        if(mc.thePlayer.isSneaking())
            return;

        final SpeedMode speedMode = getMode();

        if(speedMode != null)
            speedMode.onMove(event);
    }

    @EventTarget
    public void onTick(final TickEvent event) {
        if(mc.thePlayer.isSneaking())
            return;

        final SpeedMode speedMode = getMode();

        if(speedMode != null)
            speedMode.onTick();
    }

    @Override
    public void onEnable() {
        if(mc.thePlayer == null)
            return;

        mc.timer.timerSpeed = 1F;

        final SpeedMode speedMode = getMode();

        if(speedMode != null)
            speedMode.onEnable();
    }

    @Override
    public void onDisable() {
        if(mc.thePlayer == null)
            return;

        mc.timer.timerSpeed = 1F;

        final SpeedMode speedMode = getMode();

        if(speedMode != null)
            speedMode.onDisable();
    }

    @Override
    public String getTag() {
        return modeValue.get();
    }

    private SpeedMode getMode() {
        final String mode = modeValue.get();

        for(final SpeedMode speedMode : speedModes)
            if(speedMode.modeName.equalsIgnoreCase(mode))
                return speedMode;

        return null;
    }

    private String[] getModes() {
        final List<String> list = new ArrayList<>();
        for(final SpeedMode speedMode : speedModes)
            list.add(speedMode.modeName);
        return list.toArray(new String[0]);
    }
}

File: shared/main/java/net/ccbluex/liquidbounce/features/module/modules/movement/Speed.kt
/*
 * LiquidBounce Hacked Client
 * A free open source mixin-based injection hacked client for Minecraft using Minecraft Forge.
 * https://github.com/CCBlueX/LiquidBounce/
 */
package net.ccbluex.liquidbounce.features.module.modules.movement

import net.ccbluex.liquidbounce.LiquidBounce
import net.ccbluex.liquidbounce.event.*
import net.ccbluex.liquidbounce.features.module.Module
import net.ccbluex.liquidbounce.features.module.ModuleCategory
import net.ccbluex.liquidbounce.features.module.ModuleInfo
import net.ccbluex.liquidbounce.features.module.modules.movement.speeds.SpeedMode
import net.ccbluex.liquidbounce.features.module.modules.movement.speeds.aac.*
import net.ccbluex.liquidbounce.features.module.modules.movement.speeds.minorACs.*
import net.ccbluex.liquidbounce.features.module.modules.movement.speeds.ncp.*
import net.ccbluex.liquidbounce.features.module.modules.movement.speeds.other.*
import net.ccbluex.liquidbounce.features.module.modules.movement.speeds.spartan.SpartanYPort
import net.ccbluex.liquidbounce.features.module.modules.movement.speeds.spectre.SpectreBHop
import net.ccbluex.liquidbounce.features.module.modules.movement.speeds.spectre.SpectreLowHop
import net.ccbluex.liquidbounce.features.module.modules.movement.speeds.spectre.SpectreOnGround
import net.ccbluex.liquidbounce.features.module.modules.movement.speeds.vanilla.Vanilla
import net.ccbluex.liquidbounce.ui.client.hud.element.elements.NotificationIcon
import net.ccbluex.liquidbounce.utils.MovementUtils
import net.ccbluex.liquidbounce.value.*

@ModuleInfo(name = "Speed", description = "Allows you to move faster.", category = ModuleCategory.MOVEMENT)
object Speed : Module()
{
	private val speedModes = arrayOf(

		// Vanilla
		Vanilla(),

		// NCP
		OldNCPBHop(), NCPFHop(), SNCPBHop(), NCPBHop(), YPort(), YPort2(), NCPYPort(), Boost(), Frame(), MiJump(), OnGround(),

		// AAC BHop, LowHop, (Y)Port, Ground speeds
		AAC1_9_10BHop(), AAC3_0_3LowHop(), AAC3_0_5BHop(), AAC3_1_0YPort(), AAC3_1_5LowHop(), AAC3_1_5FastLowHop(), AAC3_2_2LowHop(), AAC3_2_2YPort(), AAC3_3_7FlagBHop(), AAC3_3_9BHop(), AAC3_3_11LowHop(), AAC3_3_11BHop(), AAC3_3_11Ground(), AAC3_3_11Ground2(), AAC3_3_13LowHop(), AAC3_5_0BHop(), AAC3_5_0LowHop(), AAC4_4_0BHop(), AAC5BHop(),

		// AAC
		AACPort(),

		// Spartan
		SpartanYPort(),

		// Other AntiCheats
		ACP(), ACPBHop(), ACRBHop(), PACBHop(), DaedalusAACBHop(), MatrixBHop(),

		// Spectre
		SpectreLowHop(), SpectreBHop(), SpectreOnGround(), TeleportCubeCraft(),

		// Server
		HiveHop(), HypixelHop(),

		// Mineplex
		MineplexGround(), MineplexBHop(), MineplexBHop2(),

		// RedeSky
		RedeSkyBoostHop(), RedeSkySlowHop(),

		// Other
		SlowHop(), CustomSpeed()

	)

	private val speedModeMap = mapOf(*speedModes.map { it.modeName to it }.toTypedArray())

	val modeValue: ListValue = object : ListValue("Mode", speedModeMap.keys.toTypedArray(), "NCPBHop")
	{
		override fun onChange(oldValue: String, newValue: String)
		{
			if (state) onDisable()
		}

		override fun onChanged(oldValue: String, newValue: String)
		{
			if (state) onEnable()
		}
	}

	// Vanilla Speed
	val vanillaSpeedValue = object : FloatValue("VanillaSpeed", 0.5f, 0.2f, 10.0f)
	{
		override fun showCondition() = modeValue.get().equals("Vanilla", ignoreCase = true)
	}

	// Custom Speed
	private val customGroup = object : ValueGroup("Custom")
	{
		override fun showCondition() = modeValue.get().equals("Custom", ignoreCase = true)
	}
	val customSpeedValue = FloatValue("CustomSpeed", 1.6f, 0.2f, 2f)
	val customYValue = FloatValue("CustomY", 0f, 0f, 4f)
	val customTimerValue = FloatValue("CustomTimer", 1f, 0.1f, 2f)
	val customStrafeValue = BoolValue("CustomStrafe", true)
	val customResetXZValue = BoolValue("CustomResetXZ", false)
	val customResetYValue = BoolValue("CustomResetY", false)

	// AAC Port length
	val portMax = object : FloatValue("AAC-PortLength", 1f, 1f, 20f)
	{
		override fun showCondition() = modeValue.get().equals("AACPort", ignoreCase = true)
	}

	// AAC Ground timer
	val aacGroundTimerValue = object : FloatValue("AACGround-Timer", 3f, 1.1f, 10f)
	{
		override fun showCondition() = modeValue.get().equals("AAC3.3.11-Ground", ignoreCase = true) || modeValue.get().equals("AAC3.3.11-Ground2", ignoreCase = true)
	}

	// Cubecraft Port length
	val cubecraftPortLengthValue = object : FloatValue("CubeCraft-PortLength", 1f, 0.1f, 2f)
	{
		override fun showCondition() = modeValue.get().equals("CubeCraft", ignoreCase = true)
	}

	// Mineplex Ground speed
	val mineplexGroundSpeedValue = object : FloatValue("MineplexGround-Speed", 0.5f, 0.1f, 1f)
	{
		override fun showCondition() = modeValue.get().equals("Mineplex-Ground", ignoreCase = true)
	}

	// Slowhop multiplier
	val slowHopMultiplierValue = object : FloatValue("SlowHop-Multiplier", 1.011f, 1.001f, 1.015f)
	{
		override fun showCondition() = modeValue.get().equals("SlowHop", ignoreCase = true)
	}

	private val ncphopGroup = object : ValueGroup("NCPHop")
	{
		override fun showCondition() = modeValue.get().equals("NCPBHop", ignoreCase = true) || modeValue.get().equals("NCPFHop", ignoreCase = true)
	}
	val ncphopBoostTicks = IntegerValue("BoostTicks", 1, 1, 5)
	val ncphopNoBoostTicks = IntegerValue("NoBoostTicks", 0, 0, 5)

	private val disableOnFlagValue = BoolValue("DisableOnFlag", true)

	private var mode: SpeedMode? = null

	override val tag: String
		get() = modeValue.get()

	init
	{
		ncphopGroup.addAll(ncphopBoostTicks, ncphopNoBoostTicks)

		customGroup.addAll(customSpeedValue, customYValue, customTimerValue, customStrafeValue, customResetXZValue, customResetYValue)
	}

	@EventTarget
	fun onUpdate(@Suppress("UNUSED_PARAMETER") event: UpdateEvent?)
	{
		val thePlayer = mc.thePlayer ?: return

		if (thePlayer.sneaking) return

		if (MovementUtils.isMoving(thePlayer)) thePlayer.sprinting = true

		mode?.onUpdate()
	}

	@EventTarget
	fun onMotion(event: MotionEvent)
	{
		val thePlayer = mc.thePlayer ?: return

		if (thePlayer.sneaking) return

		mode?.onMotion(event.eventState)
	}

	@EventTarget
	fun onMove(event: MoveEvent)
	{
		if ((mc.thePlayer ?: return).sneaking) return
		mode?.onMove(event)
	}

	@EventTarget
	fun onTick(@Suppress("UNUSED_PARAMETER") event: TickEvent?)
	{
		if ((mc.thePlayer ?: return).sneaking) return

		mode?.onTick()
	}

	@EventTarget
	fun onPacket(event: PacketEvent)
	{
		if (classProvider.isSPacketPlayerPosLook(event.packet) && disableOnFlagValue.get())
		{
			val thePlayer = mc.thePlayer ?: return

			state = false

			MovementUtils.zeroXZ(thePlayer)
			thePlayer.jumpMovementFactor = 0.02F

			LiquidBounce.hud.addNotification(NotificationIcon.WARNING_RED, "Speed", "A teleport has been detected. Disabled Speed to prevent more flags.", 1000L)
		}
	}

	override fun onEnable()
	{
		mode = speedModeMap[modeValue.get()]

		mc.thePlayer ?: return

		mc.timer.timerSpeed = 1f

		mode?.onEnable()
	}

	override fun onDisable()
	{
		mc.thePlayer ?: return

		mc.timer.timerSpeed = 1f

		mode?.onDisable()
	}

	fun allowSprintBoost(): Boolean
	{
		val mode = modeValue.get()
		return sequenceOf("AAC3.3.11-Ground", "AAC3.3.11-Ground2", "AACPort", "ACP").any { mode.equals(it, ignoreCase = true) }
	}
}
